cluster/watchdog.go

```diff
@@ -0,0 +1,101 @@
+package cluster
+
+import (
+	"sync"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+// Watchdog listen to cluster events ans handle container rescheduling
+type Watchdog struct {
+	sync.Mutex
+	cluster Cluster
+}
+
+// Handle cluster callbacks
+func (w *Watchdog) Handle(e *Event) error {
+	switch e.Status {
+	case "engine_reconnect":
+		go w.removeDuplicateContainers(e.Engine)
+	case "engine_disconnect":
+		go w.rescheduleContainers(e.Engine)
+	case "die", "destroy", "kill", "oom", "start", "stop", "rename":
+		go w.reschedulePendingContainers()
+	}
+
+	return nil
+}
+
+// Remove Duplicate containers when a node comes back
+func (w *Watchdog) removeDuplicateContainers(e *Engine) {
+	for _, container := range e.Containers() {
+		if w.cluster.Container(container.Config.SwarmID()) != nil {
+			// container already exists in the cluster, destroy it
+			e.RemoveContainer(container, true, true)
+		}
+	}
+}
+
+// Try to reschedule containers if possible
+func (w *Watchdog) reschedulePendingContainers() {
```

Comment:
There is a type `pendingContainer` in cluster/swarm/cluster.go. I think it better to differentiate the 2 concepts.